---
title: "scrub lostruct"
author: "Devon DeRaad"
date: "7/15/2020"
output: html_document
---

```{r, results = "hide", warning=FALSE}
library(lostruct)
library(vcfR)
library(adegenet)
library(gaston)
library(ggplot2)
library(gridExtra)
```

```{r}
#read in vcf as vcfR
vcfR <- read.vcfR("~/Desktop/aph.data/unzipped.filtered.vcf")
dim(vcfR@gt)
#convert to genlight
genlight <- vcfR2genlight(vcfR)

#create list of chromosomes we want to search
chrom.tab<-table(genlight$chromosome)
chrom.tab<-chrom.tab[chrom.tab >100] #retain scaffolds with greater than 100 SNPs
#retain only scaffolds mapped to chromosomes and reorder to match genomic position
chrom.tab<-chrom.tab[c(1,7,8,9,10,11,12,13,14,15,16,2,3,4,5,6,17)]
#list the ordered names of the chromosomes we want to search
chroms<-names(chrom.tab)
```

```{r}
chrom.list<-list() #initialize empty list to hold chroms
snp.list<-list() #init empty list to hold snps
#loop to create a matrix for each chromosome we want to search
for (i in chroms){
  chrom.mat<-t(as.matrix(genlight))
  chrom.list[[i]]<-chrom.mat[genlight@chromosome== i,]
  snp.list[[i]]<-genlight@position[genlight@chromosome == i]
}

#create a df that holds the starting and ending position of each window we will create
#loop to create a matrix for each chromosome we want to search
start.window<-c()
end.window<-c()
chrom.window<-c()
window.df<-data.frame()
for (i in chroms){
  chrom<-as.vector(droplevels(genlight@chromosome[genlight@chromosome == i]))
  pos<-genlight@position[genlight@chromosome == i]
  for (j in 1:floor(length(pos)/50)){
    start.window[j]<-pos[(50*j)-49]
    end.window[j]<-pos[(50*j)]
    chrom.window[j]<-chrom[50*j]
  }
  window.df<-rbind(window.df,cbind(chrom.window,start.window,end.window))
  start.window<-c() #empty vectors
  end.window<-c()
  chrom.window<-c()
}

#window.df shows the start and end pos of all 700 windows

#loop to make 50 SNP windowed pca for each chromosome individually
pcs<-list() #open empty list
for (i in 1:length(chrom.list)){
  pcs[[i]] <- eigen_windows(chrom.list[[i]], win=50, k=2)
}

#loop to combine all 50 SNP windows into a single coherent df
#create chrom.holder to track the chromosome of each window
df<-pcs[[1]] #init df
chrom.holder<- rep(chroms[1], times=nrow(pcs[[1]])) #init chrom holder
for (i in 2:length(pcs)){
  df<-rbind(df,pcs[[i]])
  chrom.holder<-c(chrom.holder, rep(chroms[i], times=nrow(pcs[[i]])))
}

#calculate pcdist between all windows
pcdist <- pc_dist(df,npc=2)

#check out pcdist
dim(pcdist)
#make a vector to index out NA values
na.inds <- is.na(pcdist[,1])
table(na.inds)
#index NAs out of pcdist, and also out of window.df
pcdist<-pcdist[!na.inds,!na.inds]
window.df<-window.df[!na.inds,]

dim(pcdist)
dim(window.df) #num of rows should match

#calculate mds for all windows
mds <- cmdscale(pcdist, eig=TRUE, k=2)
```

```{r}
#
plot(mds$points, xlab="Coordinate 1", ylab="Coordinate 2", col=as.factor(window.df$chrom.window))
mincirc <- lostruct:::enclosing_circle(mds$points[,1:2] )
mds.corners <- corners( mds$points[,1:2], prop=.05 )
corner.cols <- c("red","blue","purple")
ccols <- rep("black",nrow(mds$points))
for (k in 1:ncol(mds.corners)) {
  ccols[ mds.corners[,k] ] <- corner.cols[k]
}
plot( mds$points[,1:2], pch=20, col=adjustcolor(ccols,0.75), 
      xlab="MDS coordinate 1", ylab="MDS coordinate 2", 
      xlim=mincirc$ctr[1]+c(-1,1)*mincirc$rad,
      ylim=mincirc$ctr[2]+c(-1,1)*mincirc$rad )
#plot_circle( mincirc, col='red' )
#points( mincirc$three, col='red', cex=2 )
#points( mds$points[mincirc$index,], col='red', cex=1.5 )
```

```{r}
#axis 1
ggplot(as.data.frame(mds$points), aes(x=1:nrow(mds$points), y=(mds$points[,1]),color=window.df$chrom.window))+
  geom_point()+
  scale_color_manual(values = rep(c("black","grey"), times=10))+
  theme_classic()+
  theme(legend.position = 'none')

#axis 2
ggplot(as.data.frame(mds$points), aes(x=1:nrow(mds$points), y=(mds$points[,2]),color=window.df$chrom.window))+
  geom_point()+
  scale_color_manual(values = rep(c("black","grey"), times=10))+
  theme_classic()+
  theme(legend.position = 'none')

#pull out outlier points
y<-as.data.frame(mds$points)[mds$points[,1] > .25,][,1]
x<-(1:nrow(mds$points))[mds$points[,1] > .25]
outlier.values<-as.data.frame(cbind(x,y))

#axis 1
ggplot(as.data.frame(mds$points), aes(x=1:nrow(mds$points), y=(mds$points[,1]),color=window.df$chrom.window))+
  geom_point()+
  scale_color_manual(values = rep(c("black","grey"), times=10))+
  theme_classic()+
  theme(legend.position = 'none')+
  geom_point(data = outlier.values, aes(x=x, y=y), color = "red")

#pull out outlier points
y<-as.data.frame(mds$points)[mds$points[,2] > .25,][,2]
x<-(1:nrow(mds$points))[mds$points[,2] > .25]
outlier.values<-as.data.frame(cbind(x,y))

#axis 2
ggplot(as.data.frame(mds$points), aes(x=1:nrow(mds$points), y=(mds$points[,2]),color=window.df$chrom.window))+
  geom_point()+
  scale_color_manual(values = rep(c("black","grey"), times=10))+
  theme_classic()+
  theme(legend.position = 'none')+
  geom_point(data = outlier.values, aes(x=x, y=y), color = "red")

```

```{r}
#check out potential outlier windows
window.df[mds$points[,1] > .5,]

lostruct.10<-new("genlight", as.matrix(genlight)[,genlight@chromosome == "Pseudochr10" &
                                                      genlight@position >= 4964609 &
                                                      genlight@position <= 7385252])

lostruct.14<-new("genlight", as.matrix(genlight)[,genlight@chromosome == "Pseudochr14" &
                                                      genlight@position >= 3183105 &
                                                      genlight@position <= 6739757])


putative.inversion.list<-list(lostruct.10, lostruct.14)
names<-c("lostruct.10","lostruct.14")
par(mfrow=c(1,2))
for (i in 1:length(putative.inversion.list)){
  pca <- glPca(putative.inversion.list[[i]], nf=10)
  pca.scores<-as.data.frame(pca$scores)
  plot(pca.scores$PC1,pca.scores$PC2, col=as.factor(substr(putative.inversion.list[[i]]$ind.names, 1,3)),
       main=names[i])
}

dev.off()

```

```{r}
#vet each potentially inverted region by calculating heterozygosity
#in an inversion we expect 2 distinct classes of heterozygosity btwn het and homozygous individuals
lostruct.matrix<-as.matrix(lostruct.10)
#calculate heterozygosity in the inverted region
het<-c()
for (i in 1:nrow(lostruct.matrix)){
  het[i]<-sum(na.omit(lostruct.matrix[i,] == 1))/sum(!is.na(lostruct.matrix[i,]))
}
ggplot(as.data.frame(het), aes(x=het))+
  geom_histogram(bins=15,color="black", fill="grey")+
  labs(x="Heterozygosity across inverted region", y = "Frequency")+
  theme_classic()
plot(x=pca.scores$PC1, y=het)

#14
lostruct.matrix<-as.matrix(lostruct.14)
#calculate heterozygosity in the inverted region
het<-c()
for (i in 1:nrow(lostruct.matrix)){
  het[i]<-sum(na.omit(lostruct.matrix[i,] == 1))/sum(!is.na(lostruct.matrix[i,]))
}
ggplot(as.data.frame(het), aes(x=het))+
  geom_histogram(bins=15,color="black", fill="grey")+
  labs(x="Heterozygosity across inverted region", y = "Frequency")+
  theme_classic()

#neither seem to be true inversions
#try each pop separately, maybe CA and woodhouse together?

```


```{r}
#subset to only CA and woodhouse and re-search
#read in vcf as vcfR
dim(vcfR@gt)
ca.wood.vcfR<-vcfR[,substr(colnames(vcfR@gt), 3,6)== "sumi" | colnames(vcfR@gt) == "FORMAT"]
ca.wood.vcfR
#convert to genlight
genlight <- vcfR2genlight(ca.wood.vcfR)
genlight@ind.names
```

```{r}
chrom.list<-list() #initialize empty list to hold chroms
snp.list<-list() #init empty list to hold snps
#loop to create a matrix for each chromosome we want to search
for (i in chroms){
  chrom.mat<-t(as.matrix(genlight))
  chrom.list[[i]]<-chrom.mat[genlight@chromosome== i,]
  snp.list[[i]]<-genlight@position[genlight@chromosome == i]
}

#create a df that holds the starting and ending position of each window we will create
#loop to create a matrix for each chromosome we want to search
start.window<-c()
end.window<-c()
chrom.window<-c()
window.df<-data.frame()
for (i in chroms){
  chrom<-as.vector(droplevels(genlight@chromosome[genlight@chromosome == i]))
  pos<-genlight@position[genlight@chromosome == i]
  for (j in 1:floor(length(pos)/50)){
    start.window[j]<-pos[(50*j)-49]
    end.window[j]<-pos[(50*j)]
    chrom.window[j]<-chrom[50*j]
  }
  window.df<-rbind(window.df,cbind(chrom.window,start.window,end.window))
  start.window<-c() #empty vectors
  end.window<-c()
  chrom.window<-c()
}

#window.df shows the start and end pos of all 700 windows

#loop to make 50 SNP windowed pca for each chromosome individually
pcs<-list() #open empty list
for (i in 1:length(chrom.list)){
  pcs[[i]] <- eigen_windows(chrom.list[[i]], win=50, k=2)
}

#loop to combine all 50 SNP windows into a single coherent df
#create chrom.holder to track the chromosome of each window
df<-pcs[[1]] #init df
chrom.holder<- rep(chroms[1], times=nrow(pcs[[1]])) #init chrom holder
for (i in 2:length(pcs)){
  df<-rbind(df,pcs[[i]])
  chrom.holder<-c(chrom.holder, rep(chroms[i], times=nrow(pcs[[i]])))
}

#calculate pcdist between all windows
pcdist <- pc_dist(df,npc=2)

#check out pcdist
dim(pcdist)
#make a vector to index out NA values
na.inds <- is.na(pcdist[,1])
table(na.inds)
#index NAs out of pcdist, and also out of window.df
pcdist<-pcdist[!na.inds,!na.inds]
window.df<-window.df[!na.inds,]

dim(pcdist)
dim(window.df) #num of rows should match

#calculate mds for all windows
mds <- cmdscale(pcdist, eig=TRUE, k=2)

#
plot(mds$points, xlab="Coordinate 1", ylab="Coordinate 2", col=as.factor(window.df$chrom.window))
mincirc <- lostruct:::enclosing_circle(mds$points[,1:2] )
mds.corners <- corners( mds$points[,1:2], prop=.05 )
corner.cols <- c("red","blue","purple")
ccols <- rep("black",nrow(mds$points))
for (k in 1:ncol(mds.corners)) {
  ccols[ mds.corners[,k] ] <- corner.cols[k]
}
plot( mds$points[,1:2], pch=20, col=adjustcolor(ccols,0.75), 
      xlab="MDS coordinate 1", ylab="MDS coordinate 2", 
      xlim=mincirc$ctr[1]+c(-1,1)*mincirc$rad,
      ylim=mincirc$ctr[2]+c(-1,1)*mincirc$rad )
#plot_circle( mincirc, col='red' )
#points( mincirc$three, col='red', cex=2 )
#points( mds$points[mincirc$index,], col='red', cex=1.5 )

#axis 1
ggplot(as.data.frame(mds$points), aes(x=1:nrow(mds$points), y=(mds$points[,1]),color=window.df$chrom.window))+
  geom_point()+
  scale_color_manual(values = rep(c("black","grey"), times=10))+
  theme_classic()+
  theme(legend.position = 'none')

#axis 2
ggplot(as.data.frame(mds$points), aes(x=1:nrow(mds$points), y=(mds$points[,2]),color=window.df$chrom.window))+
  geom_point()+
  scale_color_manual(values = rep(c("black","grey"), times=10))+
  theme_classic()+
  theme(legend.position = 'none')

#pull out outlier points axis 1
y<-as.data.frame(mds$points)[mds$points[,1] < -.52,][,1]
x<-(1:nrow(mds$points))[mds$points[,1] < -.52]
outlier.values<-as.data.frame(cbind(x,y))

#axis 1
ggplot(as.data.frame(mds$points), aes(x=1:nrow(mds$points), y=(mds$points[,1]),color=window.df$chrom.window))+
  geom_point()+
  scale_color_manual(values = rep(c("black","grey"), times=10))+
  theme_classic()+
  theme(legend.position = 'none')+
  geom_point(data = outlier.values, aes(x=x, y=y), color = "red")

#pull out outlier points
y<-as.data.frame(mds$points)[mds$points[,2] < -.4,][,2]
x<-(1:nrow(mds$points))[mds$points[,2] < -.4]
outlier.values<-as.data.frame(cbind(x,y))

#axis 2
ggplot(as.data.frame(mds$points), aes(x=1:nrow(mds$points), y=(mds$points[,2]),color=window.df$chrom.window))+
  geom_point()+
  scale_color_manual(values = rep(c("black","grey"), times=10))+
  theme_classic()+
  theme(legend.position = 'none')+
  geom_point(data = outlier.values, aes(x=x, y=y), color = "red")

```

```{r}
#check out potential outlier windows
window.df[mds$points[,1] < -.55,]

lostruct.1A<-new("genlight", as.matrix(genlight)[,genlight@chromosome == "Pseudochr1A" &
                                                      genlight@position >= 28910820 &
                                                      genlight@position <= 31128480])

lostruct.3<-new("genlight", as.matrix(genlight)[,genlight@chromosome == "Pseudochr3" &
                                                      genlight@position >= 41997455 &
                                                      genlight@position <= 43788237])

lostruct.4<-new("genlight", as.matrix(genlight)[,genlight@chromosome == "Pseudochr4" &
                                                      genlight@position >= 21428393 &
                                                      genlight@position <= 26789511])

lostruct.Z<-new("genlight", as.matrix(genlight)[,genlight@chromosome == "PseudochrZ" &
                                                      genlight@position >= 27586708 &
                                                      genlight@position <= 32848433])



putative.inversion.list<-list(lostruct.1A, lostruct.3)
names<-c("lostruct.1A","lostruct.3")
par(mfrow=c(1,2))
for (i in 1:length(putative.inversion.list)){
  pca <- glPca(putative.inversion.list[[i]], nf=10)
  pca.scores<-as.data.frame(pca$scores)
  plot(pca.scores$PC1,pca.scores$PC2, col=as.factor(substr(putative.inversion.list[[i]]$ind.names, 1,3)),
       main=names[i])
}

dev.off()

```

```{r}
#vet each potentially inverted region by calculating heterozygosity
#in an inversion we expect 2 distinct classes of heterozygosity btwn het and homozygous individuals
lostruct.matrix<-as.matrix(lostruct.3)
#calculate heterozygosity in the inverted region
het<-c()
for (i in 1:nrow(lostruct.matrix)){
  het[i]<-sum(na.omit(lostruct.matrix[i,] == 1))/sum(!is.na(lostruct.matrix[i,]))
}
ggplot(as.data.frame(het), aes(x=het))+
  geom_histogram(bins=15,color="black", fill="grey")+
  labs(x="Heterozygosity across inverted region", y = "Frequency")+
  theme_classic()
plot(x=pca.scores$PC1, y=het)

#1A
lostruct.matrix<-as.matrix(lostruct.1A)
#calculate heterozygosity in the inverted region
het<-c()
for (i in 1:nrow(lostruct.matrix)){
  het[i]<-sum(na.omit(lostruct.matrix[i,] == 1))/sum(!is.na(lostruct.matrix[i,]))
}
ggplot(as.data.frame(het), aes(x=het))+
  geom_histogram(bins=15,color="black", fill="grey")+
  labs(x="Heterozygosity across inverted region", y = "Frequency")+
  theme_classic()

#no true inversions can be readily located
```

